ls1intum/Artemis
# Feature/Programming-Exercises/Run test cases after due date

### Checklist

### Motivation and Context



Some courses may require test cases, that are only run after the exercise due date has passed.
This way the instructors can hold tests back so that the students are not bruteforcing their way through a submission.

### Description


- Test cases can now have the flag `afterDueDate`
- When the build result comes from our CI, we identify test cases that have this flag to update the resultString (e.g. `1 of 1 passed (preliminary)` (before due date) vs `2 of 4 passed` (after due date)) and delete feedbacks for after due date tests. The score is not changed (would be both `50%` before and after the due date in the example above).

This PR also contains refactoring on the test-case-manage component to make the inline editing table fields reusable and improving the `updateScoreFromTestCases` control flow in the ResultService.

**Known limitations:**
- ~As the result with all test cases will be run _after_ the due date has passed, it will have the "ungraded" flag. I'm not sure how we should handle this.~ => https://github.com/ls1intum/Artemis/pull/712
- With the current implementation, afterDueDate test cases should not be part of the problem statement

- [x] Implement editing mode in client for changing the afterDueDate setting
- [x] Add new field afterDueDate to programming exercise test case
- [x] Refactor some of the ngx table code to make it more reusable (editable fields)
- [x] Implement functionality to update the afterDueDate in the testCaseResource
- [x] Implement result feedback filtering for test cases before the due date passes
- [x] Write tests for generic table edit components
- [x] Write tests for manage-test-cases table
- [x] Write more service tests for excluding afterDueDate test cases from calculation
- [x] Don’t show 100% but correct score, show info in result string “(preliminary)”

### Steps for Testing


1. Log in to ArTEMiS
2. Go to course administration > programming exercises > manage test cases
3. Set the afterDueDate flag for a couple of test cases
4. Give the programming exercise a due date in the future
5. Go to the code editor, submit
6. Check that the afterDueDate test cases are _not_ part of the feedback and result string (should end with "preliminary")
7. Set the programming exercise due date to the past
8. Go to the code editor, submit
9. Check that the afterDueDate test cases are now part of the feedback and result string

### Screenshots



**Set afterDueDate flag:**

![image](https://user-images.githubusercontent.com/28230611/61590491-8f6c0a80-abba-11e9-9a2c-d1a702e2ea06.png)

**Before due date runs out:**

![before-due-date](https://user-images.githubusercontent.com/28230611/62851727-5f121a80-bce7-11e9-9417-8117a66ffdbf.jpg)

**Full score before due date:**

![image](https://user-images.githubusercontent.com/28230611/62851967-3c343600-bce8-11e9-9cd9-c1ff63c8f759.png)
